File: java_solutions_1/375.guess-number-higher-or-lower-ii.java
/*
 * @lc app=leetcode id=375 lang=java
 *
 * [375] Guess Number Higher or Lower II
 */

// @lc code=start
class Solution {
    public int getMoneyAmount(int n) {
        return java.util.stream.Stream.of(0).map(_v -> new int[n + 2][n + 2]).peek(dp -> java.util.stream.IntStream.rangeClosed(2, n).forEach(len -> java.util.stream.IntStream.rangeClosed(1, n - len + 1).forEach(i -> dp[i][i + len - 1] = java.util.stream.IntStream.rangeClosed(i, i + len - 1).map(k -> k + Math.max(dp[i][k - 1], dp[k + 1][i + len - 1])).min().getAsInt()))).map(dp -> dp[1][n]).findFirst().get();
    }
}
// @lc code=end

File: kotlin_solutions/375.guess-number-higher-or-lower-ii.kt
/*
 * @lc app=leetcode id=375 lang=kotlin
 *
 * [375] Guess Number Higher or Lower II
 */

class Solution {fun getMoneyAmount(n: Int): Int = (1..n).fold(Array(n + 1) { IntArray(n + 1) }) { dp, len -> (1..n - len + 1).forEach { i -> dp[i][i + len - 1] = if (len == 1) 0 else (i until i + len).minOf { k -> k + maxOf(if (k - 1 >= i) dp[i][k - 1] else 0, if (k + 1 < i + len) dp[k + 1][i + len - 1] else 0) } }.run{dp} }[1][n]}
